shared_client.go

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:build go1.18
+// +build go1.18
+
+package dns
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// SharedClients holds a set of SharedClient instances.
+type SharedClients struct {
+	sync.Mutex
+	// clients are created and destroyed on demand, hence 'Mutex' needs to be taken.
+	clients map[string]*SharedClient
+}
+
+func NewSharedClients() *SharedClients {
+	return &SharedClients{
+		clients: make(map[string]*SharedClient),
+	}
+}
+
+// GetSharedClient gets or creates an instance of SharedClient keyed with 'key'.
+// if 'key' is an empty sting, a new client is always created and it is not actually shared.
+// The returned 'closer' must be called once the client is no longer needed.
+func (s *SharedClients) GetSharedClient(key string, conf *Client, serverAddrStr string) (client *SharedClient, closer func()) {
+	s.Lock()
+	defer s.Unlock()
+
+	if key != "" {
+		// locate client to re-use if possible.
+		client = s.clients[key]
+	}
+	if client == nil {
+		client = newSharedClient(conf, serverAddrStr)
+		if key != "" {
+			s.clients[key] = client
+		}
+	}
+	client.refcount++
+
+	return client, func() {
+		s.Lock()
+		defer s.Unlock()
+		client.refcount--
+		if client.refcount == 0 {
+			// Make client unreachable and close it's connection.
+			// Must hold the proxy mutex for this.
+			if key != "" {
+				delete(s.clients, key)
+			}
+			// connection must be closed while holding the proxy lock to avoid a race
+			// where a new client is created with the same 5-tuple before this one is
+			// closed, which could happen if the proxy lock is released before this
+			// Close call.
```

Comment:
So, the lock on the `SharedClients` struct serves two purposes:

1) protect each client's `refcount`
2) avoid binding errors because of the race mentioned in this comment

Actually, both 1) and 2) needs to be serialized only in case of concurrent calls to `GetSharedClient` for the same key/5-tuple. So, to increase parallelism, we could use an approach similar to the one in https://github.com/cilium/cilium/blob/e40fd083a503b21dd3c88e8e21e83814356a3616/daemon/cmd/fqdn.go#L687 (having a slice of mutexes and hashing the 5-tuple to get the mutex to lock).

Looking at the two critical sections protected by the lock, I think the only relevant case where we might experience an unexpected delay is the closing of a TCP connection. Might that become a problem?
OTOH the hashing + mutex will surely increase complexity even more.

Reply:
The lock on the SharedClients also protects the map.

Reply:
Right, I missed to say that I was thinking about using a `sync.Map` instead of a plain map (docs reports one of its suggested use as `(2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys`).

Reply:
Came up with this pattern:
- client's refcount is protected by it's own mutex
- collection's (global) lock is only held during collection map operations and during the new Client creation in between a failing lookup and store, but Client creation does not make any system calls
- only client's own lock is held while it is dialing OR closing (including system calls)
- this means that while waiting for the client's lock to increase it's refcount, the client may start closing, but in this case `refcount == 0` and we may discard this client and start again.

```
func (s *SharedClients) GetSharedClient(key string, conf *Client, serverAddrStr string) (client *SharedClient, closer func()) {
	for {
		// lock for s.clients access
		s.lock.Lock()
		if key != "" {
			// locate client to re-use if possible.
			client = s.clients[key]
		}
		if client == nil {
			client = newSharedClient(conf, serverAddrStr)
			if key != "" {
				s.clients[key] = client
			}
			s.lock.Unlock()
			// new client, we are done
			break
		}
		s.lock.Unlock()

		// reusing client that may start closing while we wait for its lock
		client.Lock()
		if client.refcount > 0 {
			// not closed, add our refcount
			client.refcount++
			client.Unlock()
			break
		}
		// client was closed while we waited for it's lock, discard and try again
		client.Unlock()
		client = nil
	}

	return client, func() {
		client.Lock()
		defer client.Unlock()
		client.refcount--
		if client.refcount == 0 {
			// connection close must be completed while holding the client's lock to
			// avoid a race where a new client dials using the same 5-tuple and gets a
			// bind error.
			// The client remains findable so that new users with the same key may wait
			// for this closing to be done with.
			close(client.requestsToSend)
			client.conn = nil
			client.wg.Wait()
			// Make client unreachable
			// Must take s.lock for this.
			s.lock.Lock()
			if key != "" {
				delete(s.clients, key)
			}
			s.lock.Unlock()
		}
	}
}
```

Reply:
`SharedClients`' `lock` is now private and it is only used within `GetSharedClient` function. When it is held no client locks can be taken.